### ISSUE TITLE
Suggestion: Rewrite GROBID in Python

### ISSUE THREAD

**Issue by de-code** [kermitt2/grobid#568]
Hi Patrice,
I thought I get this out there to see what your thoughts might be on this.
GROBID is doing quite a lot of things and it is of course quite impressive to get it off the ground and get many people involved.
The GROBID code base has probably grown quite a bit over the years. Without understanding every bit, I am getting the impression that a lot of it could be simpler by having more reliable machine learning models. Personally I also see a trajectory more towards the DL models. Maybe because the way I use it (with "auto-annotation"), I am generating more data.
It's not so much about DL vs Wapiti, but that using some DL already creates more complexity by having to bridge Java and Python.
I would imagine a fresh start, re-using the existing models, could make the code base more approachable and open to change.

Summary of potential advantages:
- Reduced complexity communicating with DL model
- Simpler code base
- Easier to install (Python is more likely already available)
- Easier to extend and override behaviour (Python makes it easier to monkey-patch, if necessary)
- Python seems to be the most common language for ML

Disadvantages:
- Initial effort may be larger than it seems
- Java threads are probably more efficient than Python's multi-processing (Python's multi-threading is not useful for CPU heavy tasks)

**Comment by bnewbold** [comment]
@de-code could you comment on the use of Jython with the existing codebase to make integration with DL models easier?

My feedback on this, as an institutional user, is that Java applications are easier to deploy and operate than Python applications. Python is often available, but the labor to manage versions, compatibility, packaging, dependencies, updates, and security bugs is much greater than in the Java/JVM ecosystem, particularly over longer time scales. It also seems like a toss-up in terms of maintaining code and accepting patches: the stronger typing in Java can be verbose and slower to get started with, but for small patches and additions makes it clear what the interfaces are.

**Comment by kermitt2** [comment]
@bnewbold I think you don't want to use Jython for any DL integrations, Jython does still not support python 3.*, and has a lot of other constraints.

Since end of 2018, Grobid uses JEP to integrate any pieces of python and all the Deep Learning models of [DeLFT](https://github.com/kermitt2/delft), so in particular all the relevant state of the art sequence labelling architectures (BidLSTM-CRF, BERT-CRF, ELMo, etc.) can be used in Grobid (models are in grobid-home). According to benchmarks ([slide 31](https://www.slideshare.net/FlinkForward/flink-forward-berlin-2017-dongwon-kim-predictive-maintenance-with-apache-flink), JEP is by far the most efficient way to integrate TensorFlow in Java, as compared to the existing TF Java API or python RPC serving (Jython not being a viable option). JEP is a JNI of Cython, so it's actually often even more efficient than default python interpreter. 

Some documentation on that effort is here -> https://grobid.readthedocs.io/en/latest/Deep-Learning-models/

The integration in Grobid is transparent (same input/output as CRF Wapiti for all sequence labelling tasks) and it's also possible to use DeLFT separately for experimental works with Grobid models, so to come back to the core subject of this issue, I think it's powerful and flexible, taking advantage of rapid prototyping with Python and the efficiency and robustness for production of the existing Java developments.

**Comment by kermitt2** [comment]
@de-code About a python rewriting, I think it would be a considerable amount of work - probably several years as side project to reach the functionalities, the speed and robustness of the current implementation, if doable in Python - and I don't really see any clear technical advantages given the existing JEP integration. 

Re-writing something from scratch would bring some improvements for sure, whatever the language. But I am moderately motivated by Sisyphean tasks, most of the development efforts in Grobid were directed to production aspects, having something fast and scalable. Here it would mean solving the same problems that were already solved in Java now in Python, Python having its own list of drawbacks and bringing plenty of new technical challenges. In addition, DL models, for the moment - well after 2 years of work on that, it's 1% improvement on a few models for a huge amount of additional technical complexity.  

In my humble opinion, the advantage could be to attract a larger community of contributors, which is very important for an Open Source project... but not so sure because we're doing something very niche and certainly not the sexiest part of text mining in general.

**Comment by de-code** [comment]
Thank you both for your feedback. I am trying not to make it Python vs Java.

I guess with some DL, we are currently using Java and Python, which I think is objectively more difficult to maintain and deploy than just Java or Python. Even more so, because they also need to be able to talk to each other (the JEP version in Java needs to match the JEP binary which needs to match the Python version, I believe). Fortunately some of it can be hidden with a container.

The argument is mainly depending on using some DL. With the default dataset we haven't seen as much of an improvement. On my bioRxiv (and previously eLife) dataset I can see a bigger difference. e.g. I just re-trained the header model with the following low-level f1 scores: 0.7832 (wapiti), 0.8144 (DL no word embeddings), 0.8239 (DL with glove.840B.300d). Both DL models using my features input implementation which may be improved further by what Patrice and Luca have been working. Apart from having added the additional feature input, I haven't really spent much time trying to tune the model. I've also only trained it on between 1500 and 2000 documents for now. The results might be better if I increased the number of documents. I would expect there to be more development in DL though. Maybe I am too optimistic.

I am probably underestimating the effort.

Maybe the project would attract more interest if it there were more easily accessible components. Outside science there seem to be more interest in tables and forms.
